lib/models/updateAction.js

```diff
@@ -86,6 +130,67 @@ function doIt(action, event, callback) {
         });
 }
 
+function makeTokenListenerFunc(action, event, callback) {
+    return function tokenHandlerFunc(error, token) {
+        logger.debug('tokenHandlerFunc %s %s', error, token);
+        if (error || !token) {
+            return callback(error);
+        } else {
+            logger.debug('tokenHandlerFunc retrying with', token);
+            return doRequest(action, event, token, function cbDoReqUpdAxn(error, data) {
+                callback(error, data);
+            });
+        }
+    };
+}
+
+function doItWithToken(action, event, callback) {
+    var service,
+        tenant,
+        ruleName,
+        cached,
+        newTokenListener;
+
+    service = event.service;
+    tenant = event.tenant;
+    ruleName = event.ruleName;
+    newTokenListener = makeTokenListenerFunc(action, event, callback);
+
+    cached = getCachedToken(tenant, service, ruleName);
+    if (cached.token === null) {  // First time token is generated from trust
+        cached.emitter.once(newTokenEventName, newTokenListener);
+        if (cached.generating === false) {
+            logger.debug('generating token for %s %s %s', tenant, service, ruleName);
+            generateToken(action.parameters.trust, cached);
+        }
+    } else if (cached.generating === true) { // In the middle of getting a new one
+        logger.debug('waiting token to be generated %s %s %s', tenant, service, ruleName);
+        cached.emitter.once(newTokenEventName, newTokenListener);
+    } else {
+        doRequest(action, event, cached.token, function cbDoReqUpdAxn(error, data) {
+            logger.debug('cbDoReqUpdAxn %s %s', error, data);
+            // If authorization error, we should call generateToken and addListener to retry
```

Comment:
If the access token is expired, we generate a new one and try the action again. From this point, it's the same that the other actions.